vsgenstubs4/init.py

```diff
@@ -0,0 +1,579 @@
+
+_VapourSynthMapValue = Union[
+    SingleAndSequence[int],
+    SingleAndSequence[float],
+    SingleAndSequence[DataType],
+    SingleAndSequence[vs.VideoNode],
+    SingleAndSequence[vs.VideoFrame],
+    SingleAndSequence[vs.AudioNode],
+    SingleAndSequence[vs.AudioFrame],
+    SingleAndSequence['VSMapValueCallback[Any]']
+]
+
+BoundVSMapValue = TypeVar('BoundVSMapValue', bound=_VapourSynthMapValue)
+
+VSMapValueCallback = Callable[..., BoundVSMapValue]
+
+
+vs_value_type = '_VapourSynthMapValue'
+site_package_dirname = 'vapoursynth-stubs'
+
+parser = ArgumentParser()
+parser.add_argument(
+    "plugins",
+    type=str, nargs="*",
+    help="Only generate stubs for and inject specified plugin namespaces, "
+    "append these if the stubs file already exists."
+)
+parser.add_argument(
+    "--exclude-plugin", "-r",
+    metavar="VS_EXCL_PLUGIN", action="append", help="Remove selected plugin from new stubs, "
+    "or remove from existing stubs."
+)
+parser.add_argument(
+    "--load-plugin", "-p",
+    metavar="VS_PLUGIN", action="append", help="Load non-auto-loaded VapourSynth plugin."
+)
+parser.add_argument(
+    '--avs-plugin', action='append', help='Manually load AviSynth plugin.'
+)
+parser.add_argument(
+    '--output', '-o', default='@',
+    help="Where to output the file. Can be a full file path or directory."
+    "The special value '-' means output to stdout. "
+    "The spcial value '@' will install it as a stub-package inside site-packages."
+)
+parser.add_argument(
+    '--pyi-template',
+    default=join_path(path.dirname(__file__), '_vapoursynth.part.pyi'),
+    help='Don\'t use unless you know what you are doing.'
+)
+parser.add_argument(
+    '--force', '-f',
+    action='store_true',
+    help='Force rewrite of the file.'
+)
+
+
+def indent(strings: Iterable[str], spaces: int = 4) -> str:
+    return '\n'.join(' ' * spaces + line for line in strings)
+
+
+anonymous_signature = Signature(
+    [
+        Parameter('args', Parameter.VAR_POSITIONAL, annotation=vs_value_type),
+        Parameter('kwargs', Parameter.VAR_KEYWORD, annotation=vs_value_type)
+    ],
+    return_annotation=vs.VideoNode
+)
+
+sig_excepted_errors_list = [TypeError, ValueError]
+
+if vs.__version__[0] < 60:
+    sig_excepted_errors_list.append(IndexError)
+    # it tried to delete the bound value even if the signature didn't have any
+    # eg Version functions, their plugin is bound but the function hasn't got any params
+    # https://github.com/vapoursynth/vapoursynth/pull/898
+
+
+sig_excepted_errors = tuple(sig_excepted_errors_list)
+
+types = {
+    'int', 'float', 'DataType',
+    'RawNode',
+    'VideoNode', 'AudioNode',
+    'RawFrame',
+    'VideoFrame', 'AudioFrame'
+}
+
+special_callbacks_types = {
+    'std': {
+        'ModifyFrame.selector': 'VideoFrame',
+        'FrameEval.eval': 'VideoNode',
+        'Lut.function': 'Union[int, float]',
+        'Lut2.function': 'Union[int, float]',
+    },
+    'descale': {
```

Comment:
I think this should be populated by the wider vsrepo "ecosystem".
I like this idea, but I don't think any 3rd-party plugin should be *hardcoded* into this dict.
Please find a solution that can be extended to other 3rd-party plugins without explicitly calling them out here.
Otherwise this will be an uphill battle.

Reply:
Okay, then I'll add this after my next vsrepo PR, which rewrites vsrepo.py